File: 2022-07-July-LeetCoding-Challenges/Candy.go
package main

func max(n1, n2 int) int {
    if n1 >= n2 {
        return n1
    }
    return n2
}

func candy(ratings []int) int {
    n := len(ratings)
    candies := make([]int, n)
    
    // every child must have atleast one candy
    for i := 0; i < n; i++ {
        candies[i] = 1
    }
    
    // left to right to satisfy condition
    for i := 0; i < n-1; i++ {
        if ratings[i+1] > ratings[i] {
            candies[i+1] = max(candies[i]+1, candies[i+1])
        }
    }
    
    // right to left to satisfy condition
    for i := n-2; i >= 0; i-- {
        if ratings[i] > ratings[i+1] {
            candies[i] = max(candies[i+1]+1, candies[i])
        }
    }
    
    sum := 0
    for _, count := range candies {
        sum += count
    }
    return sum
}

func main() {
	ratings := []int{1, 0, 2}
	expected := 5
	actual := candy(ratings)
	if actual != expected {
		panic("actual differs from expected")
	}
}

File: 2025-06-June-LeetCoding-Challenge/Candy.py
class Solution:
    def candy(self, ratings: list[int]) -> int:
        n = len(ratings)
        candies = [1] * n  # every child must have atleast one candy
        # first pass -> going from left to right to satisfy condition
        for i in range(0, n - 1):
            if ratings[i + 1] > ratings[i]:
                candies[i + 1] = max(candies[i + 1], candies[i] + 1)
        # second pass -> going from right to left to satisfy condition
        for i in range(n - 2, -1, -1):
            if ratings[i] > ratings[i + 1]:
                candies[i] = max(candies[i], candies[i + 1] + 1)
        return sum(candies)


def main():
    ratings = [1, 0, 2]
    assert Solution().candy(ratings) == 5

    ratings = [1, 2, 2]
    assert Solution().candy(ratings) == 4


if __name__ == '__main__':
    main()
